### PR TITLE
Chunked/iterative wavelet transform

### PR DESCRIPTION
**Pull request by VBaratham**
I also fiddled with the "--all-filters" logic

I ran with and without `--chunked` and spot checked a few random spots in the wavelet transform output - everything seemed to agree.

### CODE REVIEW

File: process_nwb/wavelet_transform.py [hunk 1]
```diff
@@ -188,20 +263,37 @@ def store_wavelet_transform(elec_series, processing, filters='rat', hg_only=True
...
-                                               abs(X_wvlt),
-                                               metric='amplitude',
-                                               source_timeseries=elec_series,
-                                               starting_time=elec_series.starting_time,
-                                               rate=rate,
-                                               description=('Wavlet: ' +
-                                                            elec_series.description))
+    if chunked:
+        if not abs_only:
+            raise NotImplementError("Can't get phase from chunked wavelet transform")
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:272]:
```suggestion
            raise NotImplementError("Phase is not implemented for chunked wavelet transform")
```

File: process_nwb/wavelet_transform.py [hunk 2]
```diff
@@ -152,8 +164,71 @@ def wavelet_transform(X, rate, filters='rat', hg_only=True, X_fft_h=None, npad=N
     return Xh, X_fft_h, cfs, sds
 
 
+class ChBandIterator(AbstractDataChunkIterator):
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:167]: What's "Ch" stand for? Chunk{ed}? I'd just expand to the whole word. Ch could also be channels.

File: process_nwb/wavelet_transform.py [hunk 3]
```diff
@@ -188,20 +263,37 @@ def store_wavelet_transform(elec_series, processing, filters='rat', hg_only=True
...
-                                               source_timeseries=elec_series,
-                                               starting_time=elec_series.starting_time,
-                                               rate=rate,
-                                               description=('Wavlet: ' +
-                                                            elec_series.description))
+    if chunked:
+        if not abs_only:
+            raise NotImplementError("Can't get phase from chunked wavelet transform")
+
+        X_wvlt_abs = ChBandIterator(X, rate, filters=filters, npad=npad, hg_only=hg_only)
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:274]: Expand Ch depending on above comment.

File: process_nwb/wavelet_transform.py [hunk 4]
```diff
@@ -60,6 +61,39 @@ def hamming(n_time, rate, min_freq, max_freq):
 
     return k
 
+def get_filterbank(filters, n_time, rate, hg_only):
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:64]: You'll need to rebase and pull in the changes around nyquist check to this function.

File: process_nwb/wavelet_transform.py [hunk 5]
```diff
@@ -152,8 +164,71 @@ def wavelet_transform(X, rate, filters='rat', hg_only=True, X_fft_h=None, npad=N
...
+
+    def __next__(self):
+        ch = self._i // self.nbands
+        band = self._i % self.nbands
+        self._i += 1
+
+        if ch >= self.X.shape[1]:
+            raise StopIteration
+
+        print("Doing channel {} band {}".format(ch, band))
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:197]: Remove this. We could implement a logger, if more verbosity is desired.

File: process_nwb/wavelet_transform.py (hunk 2 above)
Comment (JesseLivezey) [reply process_nwb/wavelet_transform.py:167]: Actually, maybe it is Channel (I'd still expand the word).

File: process_nwb/wavelet_transform.py [hunk 6]
```diff
@@ -152,8 +164,71 @@ def wavelet_transform(X, rate, filters='rat', hg_only=True, X_fft_h=None, npad=N
...
+
+        if band == 0:
+            self.X_fft_h = None
+
+        X_ch = self.X[:, [ch]]
+        data, self.X_fft_h, cfs, sds = wavelet_transform(
+            X_ch,
+            self.rate,
+            filters=[self.filterbank[band]],
+            # X_fft_h=self.X_fft_h, # Reusing X_fft_h doesn't work here because wavelet_transform() decides whether to pad X based on whether X_fft_h is passed in. Could be optimized, but this is a small price to pay for the memory gains
```
Comment (JesseLivezey) [inline process_nwb/wavelet_transform.py:207]: I'd just remove this and create a github issue to investigate whether this trick can be used.

File: scripts/preprocess_folder [hunk 7]
```diff
@@ -18,20 +18,28 @@ parser.add_argument('--frequency', type=float, default=400., help='Frequency to
 parser.add_argument('--filters', type=str, default='rat',
                     choices=['rat', 'human', 'changlab'],
                     help='Type of filter bank to use for wavelets.')
-parser.add_argument('--all_filters', action='store_false')
+parser.add_argument('--all_filters', '--all-filters', action='store_true')
```
Comment (JesseLivezey) [inline scripts/preprocess_folder:21]: I think we had talked about having the default behavior be to only store HG.

**Review by JesseLivezey — CHANGES_REQUESTED** [review]
I think this looks pretty good overall. This will need a rebase+edits and I requested a few changes.

Can you also write a test that checks that the results are the same (allclose) with and without chunking? Could just be on random data.

**Comment by JesseLivezey** [comment]
I added some utils to helping testing these type of functions

https://github.com/BouchardLab/process_nwb/blob/acf598b650bf6d2d67980d4f399c7d5ce19177c6/process_nwb/utils.py#L139
https://github.com/BouchardLab/process_nwb/blob/acf598b650bf6d2d67980d4f399c7d5ce19177c6/process_nwb/utils.py#L191

Example test that uses them here (you could create a second test that does does the wavelet step, or add another set of asserts in the existing test with this flag used.
https://github.com/BouchardLab/process_nwb/blob/master/tests/test_pipeline.py

File: process_nwb/wavelet_transform.py (hunk 2 above)
Comment (VBaratham) [reply process_nwb/wavelet_transform.py:167]: Yes, it's Channel, and I'll write the full word out

File: process_nwb/wavelet_transform.py (hunk 6 above)
Comment (VBaratham) [reply process_nwb/wavelet_transform.py:207]: > I added some utils to helping testing these type of functions
> 
> https://github.com/BouchardLab/process_nwb/blob/acf598b650bf6d2d67980d4f399c7d5ce19177c6/process_nwb/utils.py#L139
> 
> 
> https://github.com/BouchardLab/process_nwb/blob/acf598b650bf6d2d67980d4f399c7d5ce19177c6/process_nwb/utils.py#L191
> 
> Example test that uses them here (you could create a second test that does does the wavelet step, or add another set of asserts in the existing test with this flag used.
> https://github.com/BouchardLab/process_nwb/blob/master/tests/test_pipeline.py

Thanks, will do.

**Comment by JesseLivezey** [comment]
Closed by #47

**Pull request closed without merging**